src/python/WMCore/WorkQueue/WorkQueueBackend.py

```diff
@@ -458,87 +464,107 @@ def availableWork(self, thresholds, siteJobCounts, team=None,
-                                     element['RequestName'], prio, element.id, possibleSite)
-                    numElems -= 1
-                    elements.append(element)
-                    siteJobCounts.setdefault(possibleSite, {})
-                    siteJobCounts[possibleSite][prio] = siteJobCounts[possibleSite].setdefault(prio, 0) + \
-                                                        element['Jobs'] * element.get('blowupFactor', 1.0)
-                else:
-                    self.logger.debug("No available resources for %s with doc id %s",
-                                      element['RequestName'], element.id)
+            commonSites = possibleSites(element)
+            prio = element['Priority']
+            # shuffle list of common sites all the time to give everyone the same chance
+            random.shuffle(commonSites)
+            possibleSite = None
+            for site in commonSites:
+                if site in thresholds:
+                    # Count the number of jobs currently running of greater priority, if they
+                    # are less than the site thresholds, then accept this element
+                    curJobCount = sum([x[1] if x[0] >= prio else 0 for x in viewitems(siteJobCounts.get(site, {}))])
+                    self.logger.debug("Job Count: %s, site: %s thresholds: %s",
+                                      curJobCount, site, thresholds[site])
+                    if curJobCount < thresholds[site]:
+                        possibleSite = site
+                        break
+
+            if possibleSite:
+                self.logger.info("Accepting workflow: %s, with prio: %s, element id: %s, for site: %s",
+                                 element['RequestName'], prio, element.id, possibleSite)
+                numElems -= 1
+                elems.append(element)
+                siteJobCounts.setdefault(possibleSite, {})
+                siteJobCounts[possibleSite][prio] = siteJobCounts[possibleSite].setdefault(prio, 0) + \
+                                                    element['Jobs'] * element.get('blowupFactor', 1.0)
+            else:
+                self.logger.debug("No available resources for %s with doc id %s",
+                                  element['RequestName'], element.id)
 
         self.logger.info("And %d elements passed location and siteJobCounts restrictions for: %s",
-                         len(elements), self.queueUrl)
-        return elements, siteJobCounts
+                         len(elems), self.queueUrl)
+        return elems
 
     def getActiveData(self):
         """Get data items we have work in the queue for"""
         data = self.db.loadView('WorkQueue', 'activeData', {'reduce': True, 'group': True})
-        return [{'dbs_url': x['key'][0],
-                 'name': x['key'][1]} for x in data.get('rows', [])]
+        return [{'dbs_url': x['key'][0], 'name': x['key'][1]} for x in data.get('rows', [])]
```

Comment:
This line change does not contribute to the new logic. It is just a prettification of an existing line, which does not relate to the current issue and adds confusion. And I think I notice a pattern for breaking all of those similarly constructed lines down in the following  functions, obviously for better reading.